internal/command/postgres/restart.go

```diff
@@ -52,136 +57,282 @@ func runRestart(ctx context.Context) error {
+	// single node.
+	if len(machines) > 1 {
+		pgclient := flypg.New(appName, dialer)
+
+		fmt.Fprintf(io.Out, "Performing a failover\n")
+		if err := pgclient.Failover(ctx); err != nil {
+			return fmt.Errorf("failed to trigger failover %w", err)
+		}
 	}
 
-	dialer, err := agentclient.Dialer(ctx, app.Organization.Slug)
+	fmt.Fprintf(io.Out, "Attempting to restart leader\n")
+
+	pgclient := flypg.NewFromInstance(fmt.Sprintf("[%s]", leader.PrivateIP), dialer)
+
+	if err := pgclient.RestartNodePG(ctx); err != nil {
+		return fmt.Errorf("failed to restart postgres on node: %w", err)
+	}
+
+	fmt.Fprintf(io.Out, "Postgres cluster has been successfully restarted!\n")
+
+	return nil
+}
+
+func nomadSoftRestart(ctx context.Context, vms []*api.AllocationStatus) (err error) {
+	var (
+		dialer  = agent.DialerFromContext(ctx)
+		appName = app.NameFromContext(ctx)
+		io      = iostreams.FromContext(ctx)
+	)
+
+	leader, replicas, err := nomadNodeRoles(ctx, vms)
 	if err != nil {
-		return fmt.Errorf("can't build tunnel for %s: %s", app.Organization.Slug, err)
+		return fmt.Errorf("can't fetch leader: %w", err)
+	}
+	if leader == nil {
+		return fmt.Errorf("no leader found")
 	}
 
-	fmt.Fprintf(io.Out, "Restarting Postgres\n")
+	fmt.Fprintln(io.Out, "Restarting the Postgres Process")
+
+	if len(replicas) > 0 {
+		fmt.Fprintln(io.Out, "Attempting to restart replica(s)")
 
-	for lease, machine := range machines {
-		fmt.Fprintf(io.Out, " Restarting %s with lease %s\n", machine.ID, lease)
+		for _, replica := range replicas {
+			fmt.Fprintf(io.Out, " Restarting %s\n", replica.ID)
 
-		pgclient := flypg.NewFromInstance(fmt.Sprintf("[%s]", machine.PrivateIP), dialer)
+			pgclient := flypg.NewFromInstance(fmt.Sprintf("[%s]", replica.PrivateIP), dialer)
 
-		if err := pgclient.RestartNodePG(ctx); err != nil {
-			return fmt.Errorf("failed to restart postgres on node: %w", err)
+			if err := pgclient.RestartNodePG(ctx); err != nil {
+				return fmt.Errorf("failed to restart postgres on node: %w", err)
+			}
+			// wait for health checks to pass
 		}
 	}
 
-	fmt.Fprintf(io.Out, "Restart complete\n")
+	// Don't perform failover if the cluster is only running a
+	// single node.
+	if len(vms) > 1 {
+		pgclient := flypg.New(appName, dialer)
 
-	return nil
+		fmt.Fprintf(io.Out, "Performing a failover\n")
+		if err := pgclient.Failover(ctx); err != nil {
+			return fmt.Errorf("failed to trigger failover %w", err)
+		}
+	}
+
+	fmt.Fprintf(io.Out, "Attempting to restart leader\n")
+
+	pgclient := flypg.NewFromInstance(fmt.Sprintf("[%s]", leader.PrivateIP), dialer)
+
+	if err := pgclient.RestartNodePG(ctx); err != nil {
+		return fmt.Errorf("failed to restart postgres on node: %w", err)
+	}
+	fmt.Fprintf(io.Out, "Postgres cluster has been successfully restarted!\n")
+
+	return
 }
 
-func restartNomadCluster(ctx context.Context, app *api.AppCompact) (err error) {
+func machinesHardRestart(ctx context.Context, machines []*api.Machine) (err error) {
```

Comment:
We should work to compare this process with the soft restart.  If they are very similar, we could work to remove the process level restarts.